src/arena.c

```diff
@@ -293,34 +293,44 @@ arena_slab_reg_alloc_batch(edata_t *slab, const bin_info_t *bin_info,
 
 static void
 arena_large_malloc_stats_update(tsdn_t *tsdn, arena_t *arena, size_t usize) {
-	szind_t index, hindex;
-
 	cassert(config_stats);
 
+	szind_t index = sz_size2index(usize);
 	if (usize < SC_LARGE_MINCLASS) {
-		usize = SC_LARGE_MINCLASS;
+		assert(index < SC_NBINS);
+		bin_t *bin = &arena->bins[index];
+		malloc_mutex_lock(tsdn, &bin->lock);
+		bin->stats.nmalloc++;
+		malloc_mutex_unlock(tsdn, &bin->lock);
+	} else {
+		assert(index >= SC_NBINS);
+		szind_t hindex = index - SC_NBINS;
+		LOCKEDINT_MTX_LOCK(tsdn, arena->stats.mtx);
+		locked_inc_u64(tsdn, LOCKEDINT_MTX(arena->stats.mtx),
+			&arena->stats.lstats[hindex].nmalloc, 1);
+		LOCKEDINT_MTX_UNLOCK(tsdn, arena->stats.mtx);
 	}
-	index = sz_size2index(usize);
-	hindex = (index >= SC_NBINS) ? index - SC_NBINS : 0;
-
-	locked_inc_u64(tsdn, LOCKEDINT_MTX(arena->stats.mtx),
-	    &arena->stats.lstats[hindex].nmalloc, 1);
 }
 
 static void
 arena_large_dalloc_stats_update(tsdn_t *tsdn, arena_t *arena, size_t usize) {
-	szind_t index, hindex;
-
 	cassert(config_stats);
 
+	szind_t index = sz_size2index(usize);
 	if (usize < SC_LARGE_MINCLASS) {
-		usize = SC_LARGE_MINCLASS;
+		assert(index < SC_NBINS);
+		bin_t *bin = &arena->bins[index];
```

Comment:
We'll need to use `arena_get_bin` like here (with shard = 0):

https://github.com/jemalloc/jemalloc/blob/ea5b7bea3144cd26a63510016d778eab3ca58822/src/arena.c#L1391C1-L1391C1

That being said, it's better to rename arena_t.bins to something like `bins_use_arena_get_bin` (I can do that in a separate PR) since otherwise it's easy to miss the shard index.

Reply:
> That being said, it's better to rename arena_t.bins to something like bins_use_arena_get_bin

We can do even better. I can add the appropriate autoconf detection so that we can use [the deprecated attribute](https://clang.llvm.org/docs/AttributeReference.html#deprecated). If we declare `arena.bins` like this:
```c
struct arena_s {
    ...
    bin_t __attribute__((deprecated("Do not access this field directly, use `arena_get_bin` instead"))) bins[0];
};
```

Then when you compile, anywhere you use the field directly (and don't explicitly suppress the warning) you get a compiler warning message like this:
```
src/arena.c:303:24: warning: 'bins' is deprecated: Do not access this field directly, use `arena_get_bin` instead [-Wdeprecated-declarations]
```